File: Pictok/Game/EndlessSession.swift
import Foundation
import Observation

@Observable
final class EndlessSession {
    private static let maxHearts = 5
    private static let recentIdsBufferSize = 5

    private let allPuzzles: [Puzzle]
    private let store: UserStateStore
    private let today: String
    private let selector: EndlessSelector

    private(set) var currentPuzzle: Puzzle?
    private(set) var hearts: Int = maxHearts
    private(set) var correctGuesses: Set<Character> = []
    private(set) var wrongGuesses: Set<Character> = []
    private(set) var isSolved: Bool = false
    private(set) var isFailed: Bool = false
    private(set) var hintUsedThisPuzzle: Bool = false
    private(set) var hasShownOneChanceWarning: Bool = false

    /// True when every letter in the answer has been correctly guessed (or revealed
    /// via hint), but the player has not yet tapped Submit. Win celebration is gated
    /// on the player tapping Submit, which calls `submit()`.
    var needsSubmit: Bool {
        guard let puzzle = currentPuzzle, !isSolved, !isFailed else { return false }
        return GameEngine.isSolved(answer: puzzle.answer,
                                   correctGuesses: correctGuesses,
                                   revealedLetter: nil)
    }

    init(allPuzzles: [Puzzle], store: UserStateStore, today: String,
         selector: EndlessSelector = EndlessSelector()) {
        self.allPuzzles = allPuzzles
        self.store = store
        self.today = today
        self.selector = selector
        self.currentPuzzle = selector.nextPuzzle(allPuzzles: allPuzzles,
                                                 state: store.state,
                                                 today: today)
    }

    func guess(letter: Character) {
        guard let puzzle = currentPuzzle, !isSolved, !isFailed else { return }
        let upper = Character(String(letter).uppercased())
        if correctGuesses.contains(upper) || wrongGuesses.contains(upper) { return }

        if GameEngine.isCorrect(letter: upper, in: puzzle) {
            correctGuesses.insert(upper)
            // NOTE: do not flip isSolved here. The player must tap Submit;
            // that path runs in submit().
        } else {
            wrongGuesses.insert(upper)
            hearts -= 1
            if hearts == 1 && !hasShownOneChanceWarning {
                hasShownOneChanceWarning = true
            }
            if GameEngine.isFailed(lives: hearts) {
                isFailed = true
                recordFail(id: puzzle.id)
            }
        }
    }

    func useHint() {
        guard !hintUsedThisPuzzle,
              let puzzle = currentPuzzle,
              !isSolved, !isFailed else { return }
        // Pick the first unguessed letter in the answer, left to right.
        guard let toReveal = puzzle.answer.first(where: {
            $0.isLetter && !correctGuesses.contains($0)
        }) else { return }
        correctGuesses.insert(toReveal)
        hintUsedThisPuzzle = true
        // Like guess(): do not auto-solve. needsSubmit becomes true via its
        // computed getter, and the view shows the Submit button.
    }

    /// Player-tap finisher: flips isSolved and records the solve. Only does work
    /// if the puzzle is currently `needsSubmit` (all letters revealed but not yet
    /// celebrated). Safe to call redundantly.
    func submit() {
        guard needsSubmit, let puzzle = currentPuzzle else { return }
        isSolved = true
        recordSolve(id: puzzle.id)
    }

    func advance() {
        if let prevId = currentPuzzle?.id {
            var buffer = store.state.recentEndlessIds
            buffer.append(prevId)
            if buffer.count > Self.recentIdsBufferSize {
                buffer.removeFirst(buffer.count - Self.recentIdsBufferSize)
            }
            store.state.recentEndlessIds = buffer
            store.save()
        }
        hearts = Self.maxHearts
        correctGuesses = []
        wrongGuesses = []
        isSolved = false
        isFailed = false
        hintUsedThisPuzzle = false
        hasShownOneChanceWarning = false
        currentPuzzle = selector.nextPuzzle(allPuzzles: allPuzzles,
                                            state: store.state,
                                            today: today)
    }

    private func recordSolve(id: String) {
        store.state.solvedPuzzleIds.insert(id)
        store.state.lifetimeSolvedCount += 1
        store.save()
    }

    private func recordFail(id: String) {
        store.state.failedPuzzleIds.insert(id)
        store.save()
    }
}

File: web/js/endless-session.js
import * as engine from './game-engine.js';
import { nextPuzzle } from './endless-selector.js';
import * as us from './user-state.js';

const MAX_HEARTS = 5;
const RECENT_BUFFER = 5;

export function createEndlessSession(allPuzzles, state, today, storage) {
  let puzzle = nextPuzzle(allPuzzles, state, today);

  const session = {
    get currentPuzzle() { return puzzle; },
    hearts: MAX_HEARTS,
    correct: new Set(),
    wrong: new Set(),
    solved: false,
    failed: false,
    hintUsed: false,
    hasShownOneChanceWarning: false,
    solvedThisSession: 0,
    get needsSubmit() {
      if (!puzzle || session.solved || session.failed) return false;
      return engine.isSolved(puzzle.answer, session.correct, null);
    },
    guess(letter) {
      if (!puzzle || session.solved || session.failed) return;
      const u = String(letter).toUpperCase();
      if (session.correct.has(u) || session.wrong.has(u)) return;
      if (engine.isCorrect(u, puzzle.answer)) {
        session.correct.add(u);
      } else {
        session.wrong.add(u);
        session.hearts -= 1;
        if (session.hearts === 1 && !session.hasShownOneChanceWarning) {
          session.hasShownOneChanceWarning = true;
        }
        if (engine.isFailed(session.hearts)) {
          session.failed = true;
          recordFail(state, puzzle, storage);
        }
      }
    },
    useHint() {
      if (session.hintUsed || !puzzle || session.solved || session.failed) return;
      const letter = engine.letterToReveal(puzzle.answer, session.correct);
      if (!letter) return;
      session.correct.add(letter);
      session.hintUsed = true;
    },
    submit() {
      if (!session.needsSubmit) return;
      session.solved = true;
      session.solvedThisSession += 1;
      recordSolve(state, puzzle, storage);
    },
    advance() {
      if (puzzle) {
        const buf = [...(state.recentEndlessIds ?? []), puzzle.id];
        state.recentEndlessIds = buf.slice(-RECENT_BUFFER);
        us.save(state, storage);
      }
      session.hearts = MAX_HEARTS;
      session.correct = new Set();
      session.wrong = new Set();
      session.solved = false;
      session.failed = false;
      session.hintUsed = false;
      session.hasShownOneChanceWarning = false;
      puzzle = nextPuzzle(allPuzzles, state, today);
    },
  };
  return session;
}

function recordSolve(state, puzzle, storage) {
  if (!state.solvedPuzzleIds.includes(puzzle.id)) {
    state.solvedPuzzleIds = [...state.solvedPuzzleIds, puzzle.id];
  }
  state.lifetimeSolvedCount += 1;
  us.save(state, storage);
}

function recordFail(state, puzzle, storage) {
  if (!state.failedPuzzleIds.includes(puzzle.id)) {
    state.failedPuzzleIds = [...state.failedPuzzleIds, puzzle.id];
  }
  us.save(state, storage);
}
